File: website/vo2/as/assembly/vo2-formulas.ts
export function acsmVo2(speedMpm: f64, grade: f64): f64 {
  return 0.2 * speedMpm + 0.9 * speedMpm * grade + 3.5;
}

// Returns NaN on invalid input (signals "no value").
export function danielsVo2Max(velocityMpm: f64, timeMin: f64): f64 {
  if (velocityMpm < 60.0 || timeMin < 1.0) return NaN;
  const pct = 0.8
            + 0.1894393 * Math.exp(-0.012778 * timeMin)
            + 0.2989558 * Math.exp(-0.1932605 * timeMin);
  const vo2 = -4.60 + 0.182258 * velocityMpm + 0.000104 * velocityMpm * velocityMpm;
  if (pct <= 0.0 || vo2 <= 0.0) return NaN;
  return vo2 / pct;
}

export class FitnessCategory {
  constructor(public name: string, public description: string) {}
}

export function fitnessCategory(vo2Max: f64): FitnessCategory {
  if (vo2Max < 30.0) return new FitnessCategory('Poor', 'Below average cardiovascular fitness');
  if (vo2Max < 40.0) return new FitnessCategory('Fair', 'Average cardiovascular fitness');
  if (vo2Max < 50.0) return new FitnessCategory('Good', 'Above average cardiovascular fitness');
  if (vo2Max < 60.0) return new FitnessCategory('Excellent', 'High cardiovascular fitness');
  if (vo2Max < 75.0) return new FitnessCategory('Superior', 'Very high cardiovascular fitness');
  return new FitnessCategory('Elite', 'Elite-level cardiovascular fitness');
}

// Half-away-from-zero rounding, matching the Rust/Go/.NET/JS ports.
export function roundTo(v: f64, places: i32): f64 {
  const factor = Math.pow(10.0, <f64>places);
  const sign: f64 = v < 0.0 ? -1.0 : 1.0;
  return sign * Math.round(Math.abs(v) * factor) / factor;
}

@inline export function clampF(v: f64, lo: f64, hi: f64): f64 {
  return v < lo ? lo : v > hi ? hi : v;
}

@inline export function clampI(v: i32, lo: i32, hi: i32): i32 {
  return v < lo ? lo : v > hi ? hi : v;
}

// Equivalent to JS Number.prototype.toFixed — AS doesn't expose it on f64.
// Rounds half-away-from-zero to match the other ports.
export function formatFixed(v: f64, places: i32): string {
  if (isNaN(v)) return 'NaN';
  const factor = Math.pow(10.0, <f64>places);
  const sign: f64 = v < 0.0 ? -1.0 : 1.0;
  const rounded = sign * Math.round(Math.abs(v) * factor) / factor;
  if (places <= 0) return (<i64>rounded).toString();

  const negative = rounded < 0.0;
  const abs = Math.abs(rounded);
  const whole = <i64>Math.floor(abs);
  // Scale the fractional part with the same factor we used to round so we
  // recover the exact integer mantissa of the rounded value.
  const fracInt = <i64>Math.round((abs - <f64>whole) * factor);

  let frac = fracInt.toString();
  while (frac.length < places) frac = '0' + frac;
  const result = whole.toString() + '.' + frac;
  return negative ? '-' + result : result;
}

File: website/vo2/kt/src/wasmJsMain/kotlin/Vo2Formulas.kt
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sign

// ACSM Running Metabolic Equation.
// VO2 (mL/kg/min) = 0.2 × S + 0.9 × S × G + 3.5
internal fun acsmVo2(speedMpm: Double, grade: Double): Double =
    0.2 * speedMpm + 0.9 * speedMpm * grade + 3.5

internal fun danielsVo2Max(velocityMpm: Double, timeMin: Double): Double? {
    if (velocityMpm < 60.0 || timeMin < 1.0) return null
    val pct = 0.8 +
        0.1894393 * exp(-0.012778 * timeMin) +
        0.2989558 * exp(-0.1932605 * timeMin)
    val vo2 = -4.60 + 0.182258 * velocityMpm + 0.000104 * velocityMpm * velocityMpm
    if (pct <= 0.0 || vo2 <= 0.0) return null
    return vo2 / pct
}

internal data class FitnessCategory(val name: String, val description: String)

internal fun fitnessCategory(vo2Max: Double): FitnessCategory = when {
    vo2Max < 30.0 -> FitnessCategory("Poor", "Below average cardiovascular fitness")
    vo2Max < 40.0 -> FitnessCategory("Fair", "Average cardiovascular fitness")
    vo2Max < 50.0 -> FitnessCategory("Good", "Above average cardiovascular fitness")
    vo2Max < 60.0 -> FitnessCategory("Excellent", "High cardiovascular fitness")
    vo2Max < 75.0 -> FitnessCategory("Superior", "Very high cardiovascular fitness")
    else -> FitnessCategory("Elite", "Elite-level cardiovascular fitness")
}

// Half-away-from-zero rounding, matching the Rust/Go/.NET/JS/AS ports.
// Kotlin's Math.round is half-to-positive-infinity, so we sign-correct.
internal fun roundTo(v: Double, places: Int): Double {
    val factor = 10.0.pow(places)
    return sign(v) * round(abs(v) * factor) / factor
}

@Suppress("NOTHING_TO_INLINE")
internal inline fun clampF(v: Double, lo: Double, hi: Double): Double =
    if (v < lo) lo else if (v > hi) hi else v

@Suppress("NOTHING_TO_INLINE")
internal inline fun clampI(v: Int, lo: Int, hi: Int): Int =
    if (v < lo) lo else if (v > hi) hi else v

// Equivalent to JS Number.prototype.toFixed.
internal fun formatFixed(v: Double, places: Int): String {
    val factor = 10.0.pow(places)
    val rounded = sign(v) * round(abs(v) * factor) / factor
    if (places <= 0) return rounded.toLong().toString()

    val negative = rounded < 0.0
    val abs = abs(rounded)
    val whole = abs.toLong()
    val fracInt = round((abs - whole) * factor).toLong()
    val frac = fracInt.toString().padStart(places, '0')
    return (if (negative) "-" else "") + whole.toString() + "." + frac
}
